prov/efa/src/rxr/rxr_pkt_type_req.c

```diff
@@ -2498,37 +2625,67 @@ int rxr_pkt_proc_compare_rta(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
 	 */
 	if (dt == FI_INT128) {
 		for (i = 0; i < rx_entry->iov_count; ++i) {
-			ofi_int128_t src, cmp;
-			memcpy(&src, src_data + offset, sizeof(ofi_int128_t));
-			memcpy(&cmp, cmp_data + offset, sizeof(ofi_int128_t));
-
-			ofi_atomic_swap_handler(op, dt, rx_entry->iov[i].iov_base,
-									&src,
-									&cmp,
-									rx_entry->atomrsp_data + offset,
-									rx_entry->iov[i].iov_len / dtsize);
-			offset += rx_entry->iov[i].iov_len;
+			efa_mr = (struct efa_mr*) ofi_mr_map_get(&ep->util_ep.domain->mr_map, (rxr_get_rta_hdr(pkt_entry->pkt)->rma_iov + i)->key);
+			hmem_iface = efa_mr->peer.iface;
+
+			if (hmem_iface == FI_HMEM_SYSTEM) {
+				ofi_int128_t src, cmp;
+				memcpy(&src, src_data + offset, sizeof(ofi_int128_t));
+				memcpy(&cmp, cmp_data + offset, sizeof(ofi_int128_t));
+
+				ofi_atomic_swap_handler(op, dt, rx_entry->iov[i].iov_base,
+				                        &src,
+				                        &cmp,
+				                        rx_entry->atomrsp_data + offset,
+				                        rx_entry->iov[i].iov_len / dtsize);
+				offset += rx_entry->iov[i].iov_len;
+			} else {
+				FI_WARN(&rxr_prov, FI_LOG_CQ, "Atomics not supported for FI_HMEM_IFACE: %d.\n", hmem_iface);
+				return ENOSYS;
+			}
 		}
 	} else if (dt == FI_UINT128) {
 		for (i = 0; i < rx_entry->iov_count; ++i) {
-			ofi_uint128_t src, cmp;
-			memcpy(&src, src_data + offset, sizeof(ofi_uint128_t));
-			memcpy(&cmp, cmp_data + offset, sizeof(ofi_uint128_t));
-			ofi_atomic_swap_handler(op, dt, rx_entry->iov[i].iov_base,
-									&src,
-									&cmp,
-									rx_entry->atomrsp_data + offset,
-									rx_entry->iov[i].iov_len / dtsize);
-			offset += rx_entry->iov[i].iov_len;
+			efa_mr = (struct efa_mr*) ofi_mr_map_get(&ep->util_ep.domain->mr_map, (rxr_get_rta_hdr(pkt_entry->pkt)->rma_iov + i)->key);
+			hmem_iface = efa_mr->peer.iface;
+
+			if (hmem_iface == FI_HMEM_SYSTEM) {
+				ofi_uint128_t src, cmp;
+				memcpy(&src, src_data + offset, sizeof(ofi_uint128_t));
+				memcpy(&cmp, cmp_data + offset, sizeof(ofi_uint128_t));
```

Comment:
Looks like we inherited the wrong implementation.

Reply:
separate PR